libbeat/publisher/pipeline/output_test.go

```diff
@@ -0,0 +1,256 @@
+
+	"github.com/elastic/beats/v7/libbeat/common/atomic"
+	"github.com/elastic/beats/v7/libbeat/logp"
+	"github.com/elastic/beats/v7/libbeat/outputs"
+	"github.com/elastic/beats/v7/libbeat/publisher"
+	"github.com/elastic/beats/v7/libbeat/publisher/queue"
+)
+
+var (
+	SeedFlag = flag.Int64("seed", 0, "Randomization seed")
+)
+
+func TestPublish(t *testing.T) {
+	tests := map[string]func() outputs.Client{
+		"client":         newMockClient,
+		"network_client": newMockNetworkClient,
+	}
+
+	for name, ctor := range tests {
+		t.Run(name, func(t *testing.T) {
+			seedPRNG(t)
+
+			err := quick.Check(func(i uint) bool {
+				numBatches := 3000 + (i % 1000) // between 3000 and 3999
+
+				wqu := makeWorkQueue()
+				client := ctor()
+				makeClientWorker(nilObserver, wqu, client)
+
+				numEvents := atomic.MakeInt(0)
+				for batchIdx := uint(0); batchIdx <= numBatches; batchIdx++ {
+					batch := randomBatch(50, 150, wqu)
+					numEvents.Add(len(batch.Events()))
+					wqu <- batch
+				}
+
+				// Give some time for events to be published
+				timeout := time.Duration(numEvents.Load()*3) * time.Microsecond
+
+				// Make sure that all events have eventually been published
+				c := client.(interface{ Published() int })
+				return waitUntilTrue(timeout, func() bool {
+					return numEvents.Load() == c.Published()
+				})
+			}, nil)
+
+			if err != nil {
+				t.Error(err)
+			}
+		})
+	}
+}
+
+func TestPublishWithClose(t *testing.T) {
+	tests := map[string]func() outputs.Client{
+		"client":         newMockClient,
+		"network_client": newMockNetworkClient,
+	}
+
+	for name, ctor := range tests {
+		t.Run(name, func(t *testing.T) {
+			seedPRNG(t)
+
+			err := quick.Check(func(i uint) bool {
+				numBatches := 3000 + (i % 1000) // between 3000 and 3999
+
+				wqu := makeWorkQueue()
+
+				numEvents := atomic.MakeInt(0)
+
+				var wg sync.WaitGroup
+				for batchIdx := uint(0); batchIdx <= numBatches; batchIdx++ {
+					wg.Add(1)
+					go func() {
+						defer wg.Done()
+
+						batch := randomBatch(50, 150, wqu)
+						numEvents.Add(len(batch.Events()))
+						wqu <- batch
+					}()
+				}
+
+				client := ctor()
+				worker := makeClientWorker(nilObserver, wqu, client)
+
+				// Close worker before all batches have had time to be published
+				err := worker.Close()
+				require.NoError(t, err)
+
+				c := client.(interface{ Published() int })
+				remaining := numEvents.Load() - c.Published()
+				assert.Greater(t, remaining, 0)
+
+				// Start new worker to drain work queue
+				makeClientWorker(nilObserver, wqu, client)
+				wg.Wait()
+
+				// Give some time for events to be published
+				timeout := time.Duration(remaining*3) * time.Microsecond
+
+				// Make sure that all events have eventually been published
+				return waitUntilTrue(timeout, func() bool {
+					return numEvents.Load() == c.Published()
+				})
+			}, nil)
+
+			if err != nil {
+				t.Error(err)
+			}
+		})
+	}
+}
+
+const publishLatency = 5 * time.Microsecond
```

Comment:
I'm experimenting with this because the tests are consistently passing without this locally but failing in Travis CI.

Reply:
Even after trying a few different values, this did not help make the tests pass in Travis CI. So I've removed this constant and it's usages.